src/realm_dart_collections.cpp

```diff
@@ -19,84 +19,82 @@
+        : m_handle(Dart_NewFinalizableHandle_DL(handle, nullptr, 1, finalize_handle)), m_callback(callback) 
+    {}
 
-    ~Callback()
-    {
-        drop_handle();
+    ~CallbackData() {
+        delete_handle();
     }
 
-    void operator()(const realm_collection_changes_t *changes)
-    {
-        if (handle_ != nullptr)
-        {
-            Scope s; // ensure we can create handles
-            auto h = Dart_HandleFromWeakPersistent_DL(handle_);
-            // Note Dart_IsNull(h) is not exposed in DL, so we cannot check.
-            // Hence the callback has to be prepared for this eventuality.
-            callback_(h, changes);
+    void callback(const realm_collection_changes_t* changes) {
+        if (m_handle) {
+            HandleScope scope;
+            //TODO: HACK. We can not release Dart persitent handles in delete_handle on Isolate teardown since the IsolateGroup is destroyed before it.
+            //This works since Dart_WeakPersistentHandle is equivalent to Dart_FinalizableHandle. They both are FinalizablePersistentHandle internally.
+            Dart_WeakPersistentHandle weakHnd = reinterpret_cast<Dart_WeakPersistentHandle>(m_handle);
+            auto handle = Dart_HandleFromWeakPersistent_DL(weakHnd);
+            
+            //clone changes object since the Dart callback is async and changes object is valid for the duration of this method only
+            //clone failures are handled in the Dart callback
+            const realm_collection_changes_t* cloned = static_cast<realm_collection_changes_t*>(realm_clone(changes));
+            m_callback(handle, cloned);
         }
     }
+
+private:
+    //TODO: We use FinalizableHandle since it is auto-deleting. Switch to Dart_WeakPersistentHandle when the HACK is removed
+    Dart_FinalizableHandle m_handle;
+    realm_dart_on_collection_change_func_t m_callback;
 };
 
-void on_change_(void *userdata, const realm_collection_changes_t *changes)
-{
-    auto &callback = *reinterpret_cast<Callback *>(userdata);
-    callback(changes);
+void on_change_callback(void* userdata, const realm_collection_changes_t* changes) {
+    auto& callbackData = *reinterpret_cast<CallbackData*>(userdata);
+    callbackData.callback(changes);
 }
 
-void free_(void *userdata)
-{
-    auto callback = reinterpret_cast<Callback *>(userdata);
+void free_callback(void* userdata) {
+    auto callback = reinterpret_cast<CallbackData*>(userdata);
     delete callback;
 }
 
-RLM_API realm_notification_token_t *
-realm_dart_results_add_notification_callback(realm_results_t *results,
-                                             Dart_Handle userdata,
-                                             realm_dart_on_collection_change_func_t on_change,
-                                             realm_scheduler_t *scheduler)
+realm_notification_token_t* realm_dart_results_add_notification_callback(
+    realm_results_t* results, 
+    Dart_Handle notification_controller,
```

Comment:
`userdata` is a better name. This does not need to be a `NotificationController`  being parsed. Also, i think it should match the realm-core c-api function this is a front for.

Reply:
I changed the name here since this needs to be exactly notification_controller instance. The name is supposed to guide the caller. The C-API uses the name `userdata` since they don't know the context their API will be used hence a more generic term.